File: android/app/src/main/java/com/example/HaliSahaApp/data/services/BookingService.kt
package com.example.HaliSahaApp.data.services

import com.example.HaliSahaApp.data.models.*
import com.example.HaliSahaApp.data.remote.*
import com.example.HaliSahaApp.utils.AppConstants
import com.example.HaliSahaApp.utils.isWeekend
import com.example.HaliSahaApp.utils.shortFormatted
import com.google.firebase.Timestamp
import com.google.firebase.firestore.FieldValue
import com.google.gson.Gson
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.util.Calendar
import java.util.Date
import java.util.UUID

// MARK: - Booking Service
object BookingService {

    private val firebaseService = FirebaseService

    // MARK: - UI State
    private val _userBookings = MutableStateFlow<List<Booking>>(emptyList())
    val userBookings: StateFlow<List<Booking>> = _userBookings.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    // Cache
    // key: "facilityId_pitchId_date" -> List<TimeSlot>
    private val timeSlotsCache = mutableMapOf<String, List<TimeSlot>>()

    // MARK: - Fetch User Bookings
    suspend fun fetchUserBookings(): List<Booking> {
        val userId = firebaseService.currentUserId ?: throw BookingError.NotAuthenticated

        _isLoading.value = true

        return try {
            val query = firebaseService.bookingsCollection
                .whereEqualTo(FirestoreField.USER_ID, userId)
            //.orderBy(FirestoreField.CREATED_AT, com.google.firebase.firestore.Query.Direction.DESCENDING)
            // Not: İki farklı alana göre sorgu (where + order) index gerektirebilir.
            // Şimdilik client side sıralama yapabiliriz veya Firebase Console'dan index oluşturmalısın.

            val bookings: List<Booking> = firebaseService.fetchDocuments(query)

            // Client side sıralama (En yeniden eskiye)
            val sortedBookings = bookings.sortedByDescending { it.createdAt }

            _userBookings.value = sortedBookings
            _isLoading.value = false
            sortedBookings

        } catch (e: Exception) {
            _isLoading.value = false
            throw BookingError.FetchFailed(e.localizedMessage ?: "Hata")
        }
    }

    // MARK: - Fetch Upcoming Bookings
    suspend fun fetchUpcomingBookings(): List<Booking> {
        val userId = firebaseService.currentUserId ?: throw BookingError.NotAuthenticated

        val today = Calendar.getInstance().apply {
            set(Calendar.HOUR_OF_DAY, 0)
            set(Calendar.MINUTE, 0)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }.time

        val query = firebaseService.bookingsCollection
            .whereEqualTo(FirestoreField.USER_ID, userId)
            .whereGreaterThanOrEqualTo(FirestoreField.DATE, Timestamp(today))
            .whereEqualTo(FirestoreField.STATUS, BookingStatus.CONFIRMED.rawValue)

        val bookings: List<Booking> = firebaseService.fetchDocuments(query)
        return bookings.sortedBy { it.date }
    }

    // MARK: - Fetch Booking by ID
    suspend fun fetchBooking(id: String): Booking {
        return try {
            firebaseService.fetchDocument(firebaseService.bookingsCollection, id)
        } catch (e: Exception) {
            throw BookingError.NotFound
        }
    }

    // MARK: - Create Booking
    suspend fun createBooking(
        facility: Facility,
        pitch: Pitch,
        date: Date,
        startHour: Int,
        endHour: Int,
        user: User
    ): Booking {
        val userId = firebaseService.currentUserId
        val facilityId = facility.id
        val pitchId = pitch.id

        if (userId == null || facilityId == null || pitchId == null) {
            throw BookingError.NotAuthenticated
        }

        // Müsaitlik kontrolü
        val isAvailable = checkAvailability(
            facilityId = facilityId,
            pitchId = pitchId,
            date = date,
            startHour = startHour,
            endHour = endHour
        )

        if (!isAvailable) {
            throw BookingError.SlotNotAvailable
        }

        // Fiyat hesapla
        val duration = endHour - startHour
        val totalPrice = pitch.pricing.calculatePrice(
            startHour = startHour,
            duration = duration,
            isWeekend = date.isWeekend() // Extension kullanılıyor
        )
        val depositAmount = pitch.pricing.calculateDeposit(totalPrice)
        val remainingAmount = totalPrice - depositAmount

        // Booking objesi oluştur
        var booking = Booking(
            userId = userId,
            facilityId = facilityId,
            pitchId = pitchId,
            facilityName = facility.name,
            pitchName = pitch.name,
            facilityAddress = facility.address,
            facilityPhone = facility.phone,
            userFullName = user.fullName,
            userPhone = user.phone,
            date = date,
            startHour = startHour,
            endHour = endHour,
            duration = duration,
            totalPrice = totalPrice,
            depositAmount = depositAmount,
            remainingAmount = remainingAmount,
            currency = pitch.pricing.currency
        )

        // Bilet ve QR
        val ticketNumber = Booking.generateTicketNumber()
        val qrCode = generateQRCodeData(booking.copy(ticketNumber = ticketNumber))

        // Güncel veriyi tekrar oluştur (val olduğu için copy)
        val finalBooking = booking.copy(ticketNumber = ticketNumber, qrCode = qrCode)

        // Firestore'a kaydet
        val documentId = firebaseService.createDocument(
            collection = firebaseService.bookingsCollection,
            data = finalBooking
        )

        val savedBooking = finalBooking.copy(id = documentId)

        // Cache temizle
        clearTimeSlotCache(facilityId, pitchId, date)

        return savedBooking
    }

    // MARK: - Check Availability
    suspend fun checkAvailability(
        facilityId: String,
        pitchId: String,
        date: Date,
        startHour: Int,
        endHour: Int
    ): Boolean {
        val calendar = Calendar.getInstance()
        calendar.time = date
        calendar.set(Calendar.HOUR_OF_DAY, 0)
        calendar.set(Calendar.MINUTE, 0)
        val startOfDay = calendar.time

        calendar.add(Calendar.DAY_OF_YEAR, 1)
        val endOfDay = calendar.time

        val query = firebaseService.bookingsCollection
            .whereEqualTo(FirestoreField.PITCH_ID, pitchId)
            .whereGreaterThanOrEqualTo(FirestoreField.DATE, Timestamp(startOfDay))
            .whereLessThan(FirestoreField.DATE, Timestamp(endOfDay))
            .whereEqualTo(FirestoreField.STATUS, BookingStatus.CONFIRMED.rawValue)

        val existingBookings: List<Booking> = firebaseService.fetchDocuments(query)

        // Çakışma kontrolü
        for (booking in existingBookings) {
            // (start1 < end2 && end1 > start2)
            if (startHour < booking.endHour && endHour > booking.startHour) {
                return false
            }
        }
        return true
    }

    // MARK: - Get Available Time Slots
    suspend fun getAvailableTimeSlots(
        facility: Facility,
        pitch: Pitch,
        date: Date
    ): List<TimeSlot> {
        val facilityId = facility.id ?: return emptyList()
        val pitchId = pitch.id ?: return emptyList()

        // Cache kontrolü
        val cacheKey = "${facilityId}_${pitchId}_${date.shortFormatted()}"
        timeSlotsCache[cacheKey]?.let { return it }

        // Çalışma saatlerini al
        val calendar = Calendar.getInstance()
        calendar.time = date
        // Calendar.DAY_OF_WEEK: Pazar=1, Pazartesi=2...
        val dayOfWeek = calendar.get(Calendar.DAY_OF_WEEK)

        // Modelimizdeki hoursForDay fonksiyonunu kullan
        val hours = facility.operatingHours.hoursForDay(dayOfWeek)

        // Mevcut rezervasyonları çek
        calendar.set(Calendar.HOUR_OF_DAY, 0)
        calendar.set(Calendar.MINUTE, 0)
        val startOfDay = calendar.time

        calendar.add(Calendar.DAY_OF_YEAR, 1)
        val endOfDay = calendar.time

        val query = firebaseService.bookingsCollection
            .whereEqualTo(FirestoreField.PITCH_ID, pitchId)
            .whereGreaterThanOrEqualTo(FirestoreField.DATE, Timestamp(startOfDay))
            .whereLessThan(FirestoreField.DATE, Timestamp(endOfDay))
            .whereEqualTo(FirestoreField.STATUS, BookingStatus.CONFIRMED.rawValue)

        val existingBookings: List<Booking> = firebaseService.fetchDocuments(query)

        // Slot listesi oluştur
        val slots = mutableListOf<TimeSlot>()

        // "09:00" -> 9 dönüşümü
        val openHour = hours.first.take(2).toIntOrNull() ?: 9
        val closeHour = hours.second.take(2).toIntOrNull() ?: 23
        val isWeekend = date.isWeekend()

        for (hour in openHour until closeHour) {
            // Bu saatte rezervasyon var mı?
            val bookingAtHour = existingBookings.find { booking ->
                hour >= booking.startHour && hour < booking.endHour
            }
            val isBooked = bookingAtHour != null

            // Geçmiş zaman kontrolü
            val slotCalendar = Calendar.getInstance()
            slotCalendar.time = date
            slotCalendar.set(Calendar.HOUR_OF_DAY, hour)
            slotCalendar.set(Calendar.MINUTE, 0)
            val isPast = slotCalendar.time.before(Date())

            val price = pitch.pricing.calculatePrice(
                startHour = hour,
                duration = 1,
                isWeekend = isWeekend
            )

            val slot = TimeSlot(
                date = date,
                hour = hour,
                isAvailable = !isBooked && !isPast,
                bookingId = bookingAtHour?.id,
                price = price
            )
            slots.add(slot)
        }

        // Cache'e yaz
        timeSlotsCache[cacheKey] = slots
        return slots
    }

    // MARK: - Cancel Booking
    suspend fun cancelBooking(bookingId: String, reason: String? = null) {
        val userId = firebaseService.currentUserId ?: throw BookingError.NotAuthenticated

        val booking: Booking = firebaseService.fetchDocument(
            firebaseService.bookingsCollection,
            bookingId
        )

        if (booking.userId != userId) {
            throw BookingError.PermissionDenied
        }

        if (!booking.canBeCancelled) {
            throw BookingError.CannotCancel
        }

        val paymentStatus = if (booking.isRefundable) PaymentStatus.REFUNDED else PaymentStatus.PARTIAL_REFUND

        val updates = mapOf(
            FirestoreField.STATUS to BookingStatus.CANCELLED.rawValue,
            "paymentStatus" to paymentStatus.rawValue,
            "cancellationReason" to (reason ?: ""),
            FirestoreField.UPDATED_AT to FieldValue.serverTimestamp()
        )

        firebaseService.updateDocument(firebaseService.bookingsCollection, bookingId, updates)

        // Cache temizle
        clearTimeSlotCache(booking.facilityId, booking.pitchId, booking.date)
    }

    // MARK: - Process Payment (Simulation)
    suspend fun processPayment(booking: Booking, paymentMethod: PaymentMethod): PaymentResult {
        delay(1500) // 1.5 saniye bekle

        // %95 başarı oranı
        val isSuccessful = Math.random() > 0.05

        return if (isSuccessful) {
            val bookingId = booking.id ?: throw BookingError.Unknown("Rezervasyon ID yok")

            val updates = mapOf(
                FirestoreField.STATUS to BookingStatus.CONFIRMED.rawValue,
                "paymentStatus" to PaymentStatus.DEPOSIT_PAID.rawValue,
                FirestoreField.UPDATED_AT to FieldValue.serverTimestamp()
            )

            firebaseService.updateDocument(firebaseService.bookingsCollection, bookingId, updates)

            PaymentResult(true, UUID.randomUUID().toString(), "Ödeme başarıyla tamamlandı")
        } else {
            PaymentResult(false, null, "Ödeme işlemi başarısız oldu. Lütfen tekrar deneyin.")
        }
    }

    // MARK: - Generate QR Code Data (JSON String)
    private fun generateQRCodeData(booking: Booking): String {
        val map = mapOf(
            "ticketNumber" to (booking.ticketNumber),
            "date" to booking.date.toString(), // Basit string
            "startHour" to booking.startHour,
            "endHour" to booking.endHour,
            "facilityId" to booking.facilityId,
            "pitchId" to booking.pitchId
        )
        return Gson().toJson(map)
    }

    // MARK: - Cache Helpers
    private fun clearTimeSlotCache(facilityId: String, pitchId: String, date: Date) {
        val cacheKey = "${facilityId}_${pitchId}_${date.shortFormatted()}"
        timeSlotsCache.remove(cacheKey)
    }

    fun clearAllCache() {
        timeSlotsCache.clear()
    }

    // MARK: - Mock Data (Swift'teki gibi)
    fun loadMockBookings(): List<Booking> {
        val calendar = Calendar.getInstance()
        calendar.add(Calendar.DAY_OF_YEAR, 3)
        val futureDate = calendar.time

        val mockBookings = listOf(
            Booking.mockBooking,
            Booking(
                id = "mock2",
                userId = "user123",
                facilityId = "facility1",
                pitchId = "pitch1",
                facilityName = "Elit Arena",
                pitchName = "Saha 2",
                facilityAddress = "Kadıköy, İstanbul",
                facilityPhone = "+902161234567",
                userFullName = "Ahmet Yılmaz",
                userPhone = "5551234567",
                date = futureDate,
                startHour = 20,
                endHour = 21,
                totalPrice = 700.0,
                depositAmount = 140.0,
                remainingAmount = 400.0,
                currency = "TRY",
                status = BookingStatus.CONFIRMED,
                paymentStatus = PaymentStatus.DEPOSIT_PAID,
                ticketNumber = Booking.generateTicketNumber()
            )
        )

        _userBookings.value = mockBookings
        return mockBookings
    }
}

// MARK: - Enums & Data Classes

enum class PaymentMethod(val rawValue: String, val icon: String) {
    CREDIT_CARD("Kredi Kartı", "credit_card"),
    DEBIT_CARD("Banka Kartı", "credit_card"),
    WALLET("Cüzdan", "account_balance_wallet")
}

data class PaymentResult(
    val success: Boolean,
    val transactionId: String?,
    val message: String
)

sealed class BookingError(message: String) : Exception(message) {
    object NotAuthenticated : BookingError("Bu işlem için giriş yapmanız gerekiyor.")
    object NotFound : BookingError("Rezervasyon bulunamadı.")
    object SlotNotAvailable : BookingError("Seçtiğiniz saat dilimi artık müsait değil.")
    object CannotCancel : BookingError("Bu rezervasyon iptal edilemez.")
    object PermissionDenied : BookingError("Bu işlem için yetkiniz yok.")
    class FetchFailed(message: String) : BookingError("Veriler yüklenemedi: $message")
    class Unknown(message: String) : BookingError(message)
}

File: ios/HaliSahaApp/HaliSahaApp/Services/BookingService.swift
//
//  BookingService.swift
//  HaliSahaApp
//
//  Rezervasyon işlemleri servisi
//
//  Created by Mehmet Mert Mazıcı on 13.01.2026.
//

import FirebaseFirestore
import Foundation

// MARK: - Booking Service
final class BookingService: ObservableObject {

    // MARK: - Singleton
    static let shared = BookingService()

    // MARK: - Published Properties
    @Published var userBookings: [Booking] = []
    @Published var isLoading = false

    // MARK: - Private Properties
    private let firebaseService = FirebaseService.shared
    private let pendingHoldDuration: TimeInterval = 15 * 60

    // Cache
    private var timeSlotsCache: [String: [TimeSlot]] = [:]  // key: "facilityId_pitchId_date"

    // MARK: - Private Init
    private init() {}

    // MARK: - Fetch User Bookings
    @MainActor
    func fetchUserBookings() async throws -> [Booking] {
        guard let userId = firebaseService.currentUserId else {
            print("❌ fetchUserBookings: User not authenticated")
            throw BookingError.notAuthenticated
        }

        print("🔍 fetchUserBookings: Fetching for userId: \(userId)")
        isLoading = true

        do {
            let query = firebaseService.bookingsCollection
                .whereField(FirestoreField.userId, isEqualTo: userId)
                .order(by: FirestoreField.createdAt, descending: true)

            let bookings: [Booking] = try await firebaseService.fetchDocuments(query: query)
            print("✅ fetchUserBookings: Found \(bookings.count) bookings")
            self.userBookings = bookings
            isLoading = false
            return bookings

        } catch {
            print("❌ fetchUserBookings Error: \(error)")
            isLoading = false
            throw BookingError.fetchFailed(error.localizedDescription)
        }
    }

    // MARK: - Fetch Upcoming Bookings
    @MainActor
    func fetchUpcomingBookings() async throws -> [Booking] {
        guard let userId = firebaseService.currentUserId else {
            throw BookingError.notAuthenticated
        }

        let today = Calendar.current.startOfDay(for: Date())

        let query = firebaseService.bookingsCollection
            .whereField(FirestoreField.userId, isEqualTo: userId)
            .whereField(FirestoreField.date, isGreaterThanOrEqualTo: Timestamp(date: today))
            .whereField(FirestoreField.status, isEqualTo: BookingStatus.confirmed.rawValue)
            .order(by: FirestoreField.date, descending: false)

        let bookings: [Booking] = try await firebaseService.fetchDocuments(query: query)
        return bookings
    }

    // MARK: - Fetch Booking by ID
    @MainActor
    func fetchBooking(id: String) async throws -> Booking {
        do {
            let booking: Booking = try await firebaseService.fetchDocument(
                from: firebaseService.bookingsCollection,
                documentId: id
            )
            return booking
        } catch {
            throw BookingError.notFound
        }
    }

    // MARK: - Fetch Booking by Ticket Number (QR doğrulama için)
    @MainActor
    func fetchBookingByTicketNumber(_ ticketNumber: String) async throws -> Booking {
        let trimmed = ticketNumber.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { throw BookingError.notFound }

        let query = firebaseService.bookingsCollection
            .whereField("ticketNumber", isEqualTo: trimmed)
            .limit(to: 1)

        let bookings: [Booking] = try await firebaseService.fetchDocuments(query: query)
        guard let booking = bookings.first else {
            throw BookingError.notFound
        }
        return booking
    }

    // MARK: - QR İçeriğini Parse Et
    /// QR içinde JSON varsa `ticketNumber` alanını çeker; düz metin ise olduğu gibi döner.
    static func parseTicketNumber(fromScannedCode code: String) -> String? {
        let trimmed = code.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }

        // JSON dene
        if let data = trimmed.data(using: .utf8),
            let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
            let ticket = json["ticketNumber"] as? String,
            !ticket.isEmpty
        {
            return ticket
        }

        // Düz bilet numarası (HS-YYYY-XXXXXX gibi)
        return trimmed
    }

    // MARK: - Create Booking
    @MainActor
    func createBooking(
        facility: Facility,
        pitch: Pitch,
        date: Date,
        startHour: Int,
        endHour: Int,
        user: User
    ) async throws -> Booking {
        guard let userId = firebaseService.currentUserId,
            let facilityId = facility.id,
            let pitchId = pitch.id
        else {
            throw BookingError.notAuthenticated
        }

        // Müsaitlik kontrolü
        let isAvailable = try await checkAvailability(
            facilityId: facilityId,
            pitchId: pitchId,
            date: date,
            startHour: startHour,
            endHour: endHour
        )

        guard isAvailable else {
            throw BookingError.slotNotAvailable
        }

        // Fiyat hesapla
        let duration = endHour - startHour
        let totalPrice = pitch.pricing.calculatePrice(
            startHour: startHour,
            duration: duration,
            isWeekend: date.isWeekend
        )
        let depositAmount = pitch.pricing.calculateDeposit(totalPrice: totalPrice)

        let remainingAmount = totalPrice - depositAmount

        // Booking oluştur
        var booking = Booking(
            userId: userId,
            facilityId: facilityId,
            pitchId: pitchId,
            facilityName: facility.name,
            pitchName: pitch.name,
            facilityAddress: facility.address,
            facilityPhone: facility.phone,
            userFullName: user.fullName,
            userPhone: user.phone,
            date: date,
            startHour: startHour,
            endHour: endHour,
            duration: duration,
            totalPrice: totalPrice,
            depositAmount: depositAmount,
            remainingAmount: remainingAmount,
            currency: pitch.pricing.currency
        )

        // QR kod ve bilet numarası oluştur
        booking.ticketNumber = Booking.generateTicketNumber()
        booking.qrCode = generateQRCodeData(for: booking)

        // Firestore'a kaydet
        let documentId = try await firebaseService.createDocument(
            in: firebaseService.bookingsCollection,
            data: booking
        )

        booking.id = documentId

        // Cache'i temizle
        clearTimeSlotCache(facilityId: facilityId, pitchId: pitchId, date: date)

        return booking
    }

    // MARK: - Check Availability
    @MainActor
    func checkAvailability(
        facilityId: String,
        pitchId: String,
        date: Date,
        startHour: Int,
        endHour: Int
    ) async throws -> Bool {
        let startOfDay = Calendar.current.startOfDay(for: date)
        let endOfDay = Calendar.current.date(byAdding: .day, value: 1, to: startOfDay)!

        // Aynı saha, tarih ve saat aralığında aktif rezervasyon var mı?
        let query = firebaseService.bookingsCollection
            .whereField(FirestoreField.pitchId, isEqualTo: pitchId)
            .whereField(FirestoreField.date, isGreaterThanOrEqualTo: Timestamp(date: startOfDay))
            .whereField(FirestoreField.date, isLessThan: Timestamp(date: endOfDay))

        let existingBookings: [Booking] = try await firebaseService.fetchDocuments(query: query)

        // Çakışma kontrolü
        for booking in existingBookings {
            if blocksTimeSlot(booking)
                && startHour < booking.endHour
                && endHour > booking.startHour
            {
                return false
            }
        }

        return true
    }

    // MARK: - Get Available Time Slots
    @MainActor
    func getAvailableTimeSlots(
        facility: Facility,
        pitch: Pitch,
        date: Date
    ) async throws -> [TimeSlot] {
        guard let facilityId = facility.id,
            let pitchId = pitch.id
        else {
            return []
        }

        // Cache kontrolü
        let cacheKey = "\(facilityId)_\(pitchId)_\(date.shortFormatted)"
        if let cached = timeSlotsCache[cacheKey] {
            return cached
        }

        // Çalışma saatlerini al
        let dayOfWeek = Calendar.current.component(.weekday, from: date)
        let hours = facility.operatingHours.hours(for: dayOfWeek)

        // Mevcut rezervasyonları al
        let startOfDay = Calendar.current.startOfDay(for: date)
        let endOfDay = Calendar.current.date(byAdding: .day, value: 1, to: startOfDay)!

        let query = firebaseService.bookingsCollection
            .whereField(FirestoreField.pitchId, isEqualTo: pitchId)
            .whereField(FirestoreField.date, isGreaterThanOrEqualTo: Timestamp(date: startOfDay))
            .whereField(FirestoreField.date, isLessThan: Timestamp(date: endOfDay))

        let existingBookings: [Booking] = try await firebaseService.fetchDocuments(query: query)

        // Slot listesi oluştur
        var slots: [TimeSlot] = []
        let openHour = hours.open
        let closeHour = hours.close
        let isWeekend = date.isWeekend

        // 2. Metin halindeki saatleri (örn: "09:00") tam sayıya (9) çeviriyoruz
        let openHourInt = Int(hours.open.prefix(2)) ?? 9  // "09" -> 9
        // "00:00" kapanış saati gece yarısı demek, 24 olarak ele al
        var closeHourInt = Int(hours.close.prefix(2)) ?? 23  // "23" -> 23
        if closeHourInt == 0 {
            closeHourInt = 24  // Gece yarısı = 24
        }

        // Guard: Geçersiz saat aralığı kontrolü
        guard openHourInt < closeHourInt else {
            print("⚠️ Invalid hour range: open=\(openHourInt), close=\(closeHourInt)")
            return []
        }

        for hour in openHourInt..<closeHourInt {
            let isBooked = existingBookings.contains { booking in
                blocksTimeSlot(booking) && hour >= booking.startHour && hour < booking.endHour
            }

            // Geçmiş saatleri kontrol et
            let now = Date()
            let slotDate = Calendar.current.date(
                bySettingHour: hour, minute: 0, second: 0, of: date)!
            let isPast = slotDate < now

            let price = pitch.pricing.calculatePrice(
                startHour: hour,
                duration: 1,
                isWeekend: isWeekend
            )

            let slot = TimeSlot(
                date: date,
                hour: hour,
                isAvailable: !isBooked && !isPast,
                bookingId: isBooked
                    ? existingBookings.first {
                        blocksTimeSlot($0) && $0.startHour <= hour && $0.endHour > hour
                    }?.id
                    : nil,
                price: price
            )

            slots.append(slot)
        }

        // Cache'e kaydet
        timeSlotsCache[cacheKey] = slots

        return slots
    }

    // MARK: - Cancel Booking
    @MainActor
    func cancelBooking(bookingId: String, reason: String?) async throws {
        guard let userId = firebaseService.currentUserId else {
            throw BookingError.notAuthenticated
        }

        // Rezervasyonu al
        let booking: Booking = try await firebaseService.fetchDocument(
            from: firebaseService.bookingsCollection,
            documentId: bookingId
        )

        // Yetki kontrolü
        guard booking.userId == userId else {
            throw BookingError.permissionDenied
        }

        // İptal edilebilir mi kontrolü
        guard booking.canBeCancelled else {
            throw BookingError.cannotCancel
        }

        if booking.paymentStatus == .depositPaid {
            // Kapora ödenmiş — iade Cloud Function üzerinden yürütülür.
            // Sunucu hem iyzico refund'ını atar hem booking'i `cancelled`
            // durumuna geçirir ve paymentStatus'u günceller.
            _ = try await PaymentService.shared.refundDeposit(
                bookingId: bookingId,
                reason: reason
            )
        } else {
            // Ödeme alınmamış — direkt Firestore üzerinde iptal işaretle.
            // (paymentStatus alanını değiştirmiyoruz; bu sadece sunucunun yetkisinde.)
            try await firebaseService.updateDocument(
                in: firebaseService.bookingsCollection,
                documentId: bookingId,
                fields: [
                    FirestoreField.status: BookingStatus.cancelled.rawValue,
                    "cancellationReason": reason ?? "",
                    "cancelledAt": Timestamp(date: Date()),
                    FirestoreField.updatedAt: Timestamp(date: Date()),
                ]
            )
        }

        // Cache'i temizle
        clearTimeSlotCache(
            facilityId: booking.facilityId, pitchId: booking.pitchId, date: booking.date)

        // Yan etkiler: yerel hatırlatmaları sil + admin'e bildirim
        await triggerCancellationSideEffects(booking: booking)
    }

    // MARK: - Side Effects (Notifications & Reminders)

    /// Ödeme başarısı sonrası: uygunsa kullanıcının cihazında local reminder kur, admin'e bildirim yaz.
    /// iyzico akışında `BookingFlowView.handlePaymentSuccess` tarafından, Firestore'dan
    /// güncel (server tarafından `confirmed`+`depositPaid` yapılmış) booking ile çağrılır.
    @MainActor
    func triggerNewBookingSideEffects(booking: Booking, status: BookingStatus) async {
        var updatedBooking = booking
        updatedBooking.status = status
        updatedBooking.paymentStatus = .depositPaid

        // 1) Local reminder (24/2 saat öncesi) — sadece onaylı rezervasyonlarda bu cihazda
        if status == .confirmed {
            await NotificationService.shared.scheduleReminders(for: updatedBooking)
        }

        // 2) Admin'e bildirim
        guard let ownerId = await fetchFacilityOwnerId(facilityId: booking.facilityId)
        else { return }

        await AppNotificationService.shared.notify(
            AppNotification.newBookingForAdmin(adminId: ownerId, booking: updatedBooking)
        )
    }

    /// Kullanıcı iptal etti: yerel hatırlatmaları sil + admin'e bildirim.
    @MainActor
    private func triggerCancellationSideEffects(booking: Booking) async {
        // 1) Local reminder iptali
        if let id = booking.id {
            NotificationService.shared.cancelReminders(forBookingId: id)
        }

        // 2) Admin'e bildirim
        guard let ownerId = await fetchFacilityOwnerId(facilityId: booking.facilityId)
        else { return }

        await AppNotificationService.shared.notify(
            AppNotification.bookingCancelledByUser(adminId: ownerId, booking: booking)
        )
    }

    /// Tesisin sahibinin (admin) userId'sini getirir. Hata durumunda nil.
    @MainActor
    private func fetchFacilityOwnerId(facilityId: String) async -> String? {
        do {
            let facility = try await FacilityService.shared.fetchFacility(id: facilityId)
            return facility.ownerId
        } catch {
            return nil
        }
    }

    @MainActor
    private func shouldAutoConfirmBooking(facilityId: String) async -> Bool {
        // Müşteri admins/{uid}'i okuyamaz; ayar facility üzerinde mirror'lanır.
        // Facility dokümanı zaten herkese açık.
        do {
            let facility = try await FacilityService.shared.fetchFacility(id: facilityId)
            return facility.effectiveAutoConfirmBookings
        } catch {
            return true
        }
    }

    // MARK: - Generate QR Code Data
    private func generateQRCodeData(for booking: Booking) -> String {
        let data: [String: Any] = [
            "ticketNumber": booking.ticketNumber,
            "date": booking.date.ISO8601Format(),
            "startHour": booking.startHour,
            "endHour": booking.endHour,
            "facilityId": booking.facilityId,
            "pitchId": booking.pitchId,
        ]

        if let jsonData = try? JSONSerialization.data(withJSONObject: data),
            let jsonString = String(data: jsonData, encoding: .utf8)
        {
            return jsonString
        }

        return booking.ticketNumber
    }

    // MARK: - Slot Blocking
    private func blocksTimeSlot(_ booking: Booking) -> Bool {
        switch booking.status {
        case .confirmed:
            return true
        case .pending:
            return Date().timeIntervalSince(booking.createdAt) < pendingHoldDuration
        case .completed, .cancelled, .noShow:
            return false
        }
    }

    // MARK: - Clear Cache
    private func clearTimeSlotCache(facilityId: String, pitchId: String, date: Date) {
        let cacheKey = "\(facilityId)_\(pitchId)_\(date.shortFormatted)"
        timeSlotsCache.removeValue(forKey: cacheKey)
    }

    func clearAllCache() {
        timeSlotsCache.removeAll()
    }

}

// MARK: - Booking Error
enum BookingError: LocalizedError {
    case notAuthenticated
    case notFound
    case slotNotAvailable
    case cannotCancel
    case permissionDenied
    case fetchFailed(String)
    case unknown(String)

    var errorDescription: String? {
        switch self {
        case .notAuthenticated:
            return "Bu işlem için giriş yapmanız gerekiyor."
        case .notFound:
            return "Rezervasyon bulunamadı."
        case .slotNotAvailable:
            return "Seçtiğiniz saat dilimi artık müsait değil."
        case .cannotCancel:
            return "Bu rezervasyon iptal edilemez."
        case .permissionDenied:
            return "Bu işlem için yetkiniz yok."
        case .fetchFailed(let message):
            return "Veriler yüklenemedi: \(message)"
        case .unknown(let message):
            return message
        }
    }
}

